dbms/src/Compression/CompressionCodecDoubleDelta.cpp

```diff
@@ -0,0 +1,320 @@
+
+        if (double_delta == 0)
+        {
+            writer.writeBits(1, 0);
+        }
+        else
+        {
+            const auto sign = std::signbit(double_delta);
+            const auto abs_value = static_cast<typename std::make_unsigned<DeltaType>::type>(std::abs(double_delta));
+            if (double_delta > -63 && double_delta < 64)
+            {
+                writer.writeBits(2, 0b10);
+                writer.writeBits(1, sign);
+                writer.writeBits(6, abs_value);
+            }
+            else if (double_delta > -255 && double_delta < 256)
+            {
+                writer.writeBits(3, 0b110);
+                writer.writeBits(1, sign);
+                writer.writeBits(8, abs_value);
+            }
+            else if (double_delta > -2047 && double_delta < 2048)
+            {
+                writer.writeBits(4, 0b1110);
+                writer.writeBits(1, sign);
+                writer.writeBits(11, abs_value);
+            }
+            else if (double_delta > std::numeric_limits<Int32>::min() && double_delta < std::numeric_limits<Int32>::max())
+            {
+                writer.writeBits(5, 0b11110);
+                writer.writeBits(1, sign);
+                writer.writeBits(31, abs_value);
+            }
+            else
+            {
+                writer.writeBits(5, 0b11111);
+                writer.writeBits(1, sign);
+                writer.writeBits(63, abs_value);
+            }
+        }
+    }
+
+    writer.flush();
+
+    return sizeof(items_count) + sizeof(prev_value) + sizeof(prev_delta) + buffer.count();
+}
+
+template <typename T, typename DeltaType>
+void decompressDataForType(const char * source, UInt32 source_size, char * dest)
+{
+    const char * source_end = source + source_size;
+
+    const UInt32 items_count = unalignedLoad<UInt32>(source);
+    source += sizeof(items_count);
+
+    T prev_value{};
+    DeltaType prev_delta{};
+
+    if (source < source_end)
+    {
+        prev_value = unalignedLoad<T>(source);
+        unalignedStore(dest, prev_value);
+
+        source += sizeof(prev_value);
+        dest += sizeof(prev_value);
+    }
+
+    if (source < source_end)
+    {
+        prev_delta = unalignedLoad<DeltaType>(source);
+        prev_value = static_cast<DeltaType>(prev_value) + prev_delta;
+        unalignedStore(dest, prev_value);
+
+        source += sizeof(prev_delta);
+        dest += sizeof(prev_value);
+    }
+
+    ReadBufferFromMemory buffer(source, source_size - sizeof(prev_value) - sizeof(prev_delta) - sizeof(items_count));
+    BitReader reader(buffer);
+
+    // since data is tightly packed, up to 1 bit per value, and last byte is padded with zeroes,
+    // we have to keep track of items to avoid reading more that there is.
+    for (UInt32 items_read = 2; items_read < items_count && !reader.eof(); ++items_read)
+    {
+        DeltaType double_delta = 0;
+        if (reader.readBit() == 0)
+        {
+            double_delta = 0;
```

Comment:
It looks strange. What's the difference with
```
if (reader.readBit() != 0)
{
    /// all the things in lese block?
}
```

Reply:
That's a good idea.

Reply:
Done